controllers/capabilities/proxyconfigpromote_controller.go

```diff
@@ -0,0 +1,229 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	capabilitiesv1beta1 "github.com/3scale/3scale-operator/apis/capabilities/v1beta1"
+	controllerhelper "github.com/3scale/3scale-operator/pkg/controller/helper"
+	"github.com/3scale/3scale-operator/pkg/reconcilers"
+	"github.com/3scale/3scale-operator/version"
+	threescaleapi "github.com/3scale/3scale-porta-go-client/client"
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"strconv"
+)
+
+// ProxyConfigPromoteReconciler reconciles a ProxyConfigPromote object
+type ProxyConfigPromoteReconciler struct {
+	*reconcilers.BaseReconciler
+}
+
+// +kubebuilder:rbac:groups=capabilities.3scale.net,resources=proxyconfigpromotes,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=capabilities.3scale.net,resources=proxyconfigpromotes/status,verbs=get;update;patch
+
+func (r *ProxyConfigPromoteReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	reqLogger := r.Logger().WithValues("proxyconfigpromote", req.NamespacedName)
+	reqLogger.Info("Reconcile Proxy Config", "Operator version", version.Version)
+
+	proxyConfigPromote := &capabilitiesv1beta1.ProxyConfigPromote{}
+	err := r.Client().Get(r.Context(), req.NamespacedName, proxyConfigPromote)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			reqLogger.Info("resource not found. Ignoring since object must have been deleted")
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request
+		return ctrl.Result{}, err
+	}
+
+	if reqLogger.V(1).Enabled() {
+		jsonData, err := json.MarshalIndent(proxyConfigPromote, "", "  ")
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		reqLogger.V(1).Info(string(jsonData))
+	}
+
+	if proxyConfigPromote.Status.State != "Completed" {
+		status, err := r.proxyConfigPromoteReconciler(proxyConfigPromote, reqLogger, req)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		reqLogger.WithValues("status", status)
+	}
+
+	err = r.addAnnotations(proxyConfigPromote.Status.PromoteEnvironment, proxyConfigPromote.Spec.ProductCRName, proxyConfigPromote.Namespace, reqLogger)
```

Comment:
It is a very interesting idea to have in the product CR the information of the latest proxy config in staging and the latest proxy config in production. 

However, I would not add an annotation in the product CR from the proxyconfigpromote controller. Instead, I would add a watcher in the product controller to proxyconfigremote objects and update the product status with the latest information. As this is out of scope of this PR, I would remove this annotation thing and leave for another PR.

Reply:
sure will remove the function